File: clients/ios/VisualPose/Library.swift
import Foundation

final class Library {
    static let maxMs: Int64 = 120_000

    let root: URL

    init(_ root: URL) {
        self.root = root
    }

    func clipDir(_ clipId: String) -> URL {
        root.appendingPathComponent(clipId, isDirectory: true)
    }

    func mediaFile(_ meta: ClipMeta) -> URL {
        let folder = clipDir(meta.clipId)
        return meta.kind == .image
            ? folder.appendingPathComponent("clip.jpg")
            : folder.appendingPathComponent("clip.mp4")
    }

    func thumbFile(_ clipId: String) -> URL {
        clipDir(clipId).appendingPathComponent("thumb.jpg")
    }

    func metaFile(_ clipId: String) -> URL {
        clipDir(clipId).appendingPathComponent("meta.json")
    }

    func analysisFile(_ clipId: String) -> URL {
        clipDir(clipId).appendingPathComponent("analysis.json")
    }

    func stageReportFile(_ clipId: String) -> URL {
        clipDir(clipId).appendingPathComponent("stage_report.json")
    }

    func frameFeedbackFile(_ clipId: String) -> URL {
        clipDir(clipId).appendingPathComponent("frame_feedback.json")
    }

    func saveMeta(_ meta: ClipMeta) {
        Json.write(metaFile(meta.clipId), Library.toJson(meta), pretty: true)
    }

    func loadMeta(_ clipId: String) -> ClipMeta {
        let text = (try? String(contentsOf: metaFile(clipId), encoding: .utf8)) ?? "{}"
        return Library.fromJson(Json.object(from: text) ?? [:])
    }

    func listClips() -> [ClipMeta] {
        guard Json.isDirectory(root),
              let folders = try? FileManager.default.contentsOfDirectory(
                  at: root, includingPropertiesForKeys: nil, options: [.skipsHiddenFiles]
              )
        else { return [] }
        return folders.compactMap { folder in
            let path = folder.appendingPathComponent("meta.json")
            guard Json.isFile(path), let obj = Json.readObject(path) else { return nil }
            return Library.fromJson(obj)
        }
        .sorted { $0.createdAt > $1.createdAt }
    }

    func deleteClip(_ clipId: String) {
        try? FileManager.default.removeItem(at: clipDir(clipId))
    }

    @discardableResult
    func prepareReanalyze(_ clipId: String) -> ClipMeta {
        var meta = loadMeta(clipId)
        meta.status = .pending
        meta.error = nil
        meta.seedBox = nil
        meta.seeds = []
        meta.playStartMs = 0
        meta.playEndMs = nil
        saveMeta(meta)
        try? FileManager.default.removeItem(at: analysisFile(clipId))
        try? FileManager.default.removeItem(at: stageReportFile(clipId))
        try? FileManager.default.removeItem(at: frameFeedbackFile(clipId))
        return meta
    }

    func newClipId() -> String {
        UUID().uuidString.lowercased()
    }

    static func displayNameNow(_ now: Date = Date()) -> String {
        IsoClock.displayName(now)
    }

    static func createdAtIso(_ now: Date = Date()) -> String {
        IsoClock.now(now)
    }

    static func formatDurationMs(_ ms: Int, image: Bool = false) -> String {
        if image { return "--" }
        let total = max(ms / 1000, 0)
        let m = total / 60
        let s = total % 60
        return "\(m):\(String(format: "%02d", s))"
    }

    static func playWindowMs(_ meta: ClipMeta) -> (Int64, Int64) {
        let start = max(Int64(meta.playStartMs), 0)
        let end: Int64
        if let playEnd = meta.playEndMs {
            end = Int64(playEnd)
        } else if meta.durationMs > 0 {
            end = Int64(meta.durationMs)
        } else {
            end = Int64.max / 4
        }
        return (start, end <= start ? start + 1 : end)
    }

    static func statusKey(_ status: ClipStatus) -> String {
        switch status {
        case .pending: return "Pending"
        case .processing: return "Processing"
        case .done: return "Done"
        }
    }

    static func toJson(_ meta: ClipMeta) -> [String: Any] {
        [
            "clip_id": meta.clipId,
            "created_at": meta.createdAt,
            "display_name": meta.displayName,
            "duration_ms": meta.durationMs,
            "width": meta.width,
            "height": meta.height,
            "fps": meta.fps,
            "kind": meta.kind.rawValue,
            "status": meta.status.rawValue,
            "error": meta.error ?? NSNull(),
            "seed_box": meta.seedBox.map { $0.toList().map { Double($0) } as Any } ?? NSNull(),
            "seeds": meta.seeds.map { seed -> [String: Any] in
                [
                    "t_ms": seed.tMs,
                    "box": seed.box.toList().map { Double($0) },
                ]
            },
            "play_start_ms": meta.playStartMs,
            "play_end_ms": meta.playEndMs ?? NSNull(),
            "athlete_key": meta.athleteKey ?? NSNull(),
            "athlete": meta.athlete.map { $0.toJson() as Any } ?? NSNull(),
        ]
    }

    static func fromJson(_ obj: [String: Any]) -> ClipMeta {
        let seedArr = Json.optArray(obj, "seed_box")
        var seeds: [SeedMark] = []
        if let seedsArr = Json.optArray(obj, "seeds") {
            for item in seedsArr {
                guard let node = item as? [String: Any],
                      let boxArr = Json.optArray(node, "box"),
                      let box = NormBox.fromArray(boxArr)
                else { continue }
                seeds.append(SeedMark(tMs: Json.optDouble(node, "t_ms", 0), box: box))
            }
        }
        let playEnd: Int?
        if Json.isNull(obj, "play_end_ms") {
            playEnd = nil
        } else {
            playEnd = max(Int(Json.optDouble(obj, "play_end_ms", 0)), 0)
        }
        let athleteKey: String?
        if obj["athlete_key"] != nil && !Json.isNull(obj, "athlete_key") {
            let raw = Json.optString(obj, "athlete_key")
            athleteKey = raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : raw
        } else {
            athleteKey = nil
        }
        let error: String?
        if Json.isNull(obj, "error") {
            error = nil
        } else {
            let raw = Json.optString(obj, "error")
            error = raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : raw
        }
        return ClipMeta(
            clipId: Json.opt(obj, "clip_id") as? String ?? "",
            createdAt: Json.optString(obj, "created_at"),
            displayName: Json.optString(obj, "display_name"),
            durationMs: Json.optInt(obj, "duration_ms"),
            width: max(Json.optInt(obj, "width", 1), 1),
            height: max(Json.optInt(obj, "height", 1), 1),
            fps: Json.optDouble(obj, "fps", 30),
            kind: Json.optString(obj, "kind") == "image" ? .image : .video,
            status: {
                switch Json.optString(obj, "status") {
                case "processing": return .processing
                case "done": return .done
                default: return .pending
                }
            }(),
            error: error,
            seedBox: seedArr.flatMap { NormBox.fromArray($0) },
            seeds: seeds,
            playStartMs: max(Int(Json.optDouble(obj, "play_start_ms", 0)), 0),
            playEndMs: playEnd,
            athleteKey: athleteKey,
            athlete: Json.optObject(obj, "athlete").map { AthleteProfile.fromJson($0) }
        )
    }
}

File: clients/android/app/src/main/java/local/visual/corepose/Library.kt
package local.visual.corepose

import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

enum class ClipKind {
    VIDEO,
    IMAGE,
}

enum class ClipStatus {
    PENDING,
    PROCESSING,
    DONE,
}

data class NormBox(
    val x1: Float,
    val y1: Float,
    val x2: Float,
    val y2: Float,
) {
    fun toList(): List<Float> = listOf(x1, y1, x2, y2)

    companion object {
        fun fromArray(arr: JSONArray): NormBox? {
            if (arr.length() < 4) {
                return null
            }
            return NormBox(
                arr.optDouble(0).toFloat(),
                arr.optDouble(1).toFloat(),
                arr.optDouble(2).toFloat(),
                arr.optDouble(3).toFloat(),
            )
        }
    }
}

data class SeedMark(
    val tMs: Double,
    val box: NormBox,
)

data class ClipMeta(
    val clipId: String,
    val createdAt: String,
    val displayName: String,
    val durationMs: Int = 0,
    val width: Int = 1,
    val height: Int = 1,
    val fps: Double = 30.0,
    val kind: ClipKind = ClipKind.VIDEO,
    val status: ClipStatus = ClipStatus.PENDING,
    val error: String? = null,
    val seedBox: NormBox? = null,
    val seeds: List<SeedMark> = emptyList(),
    val playStartMs: Int = 0,
    val playEndMs: Int? = null,
    val athleteKey: String? = null,
    val athlete: AthleteProfile? = null,
    val scene: SceneContext? = null,
)

class Library(private val root: File) {
    fun clipDir(clipId: String): File = File(root, clipId)

    fun mediaFile(meta: ClipMeta): File {
        val folder = clipDir(meta.clipId)
        return if (meta.kind == ClipKind.IMAGE) File(folder, "clip.jpg") else File(folder, "clip.mp4")
    }

    fun thumbFile(clipId: String): File = File(clipDir(clipId), "thumb.jpg")

    fun metaFile(clipId: String): File = File(clipDir(clipId), "meta.json")

    fun analysisFile(clipId: String): File = File(clipDir(clipId), "analysis.json")

    fun stageReportFile(clipId: String): File = File(clipDir(clipId), "stage_report.json")

    fun frameFeedbackFile(clipId: String): File = File(clipDir(clipId), "frame_feedback.json")

    fun reportCorrectionFile(clipId: String): File = File(clipDir(clipId), "report_correction.json")

    fun saveMeta(meta: ClipMeta) {
        val path = metaFile(meta.clipId)
        path.parentFile?.mkdirs()
        path.writeText(toJson(meta).toString(2), Charsets.UTF_8)
    }

    fun loadMeta(clipId: String): ClipMeta {
        return fromJson(JSONObject(metaFile(clipId).readText(Charsets.UTF_8)))
    }

    fun listClips(): List<ClipMeta> {
        if (!root.isDirectory) {
            return emptyList()
        }
        return root.listFiles()
            ?.mapNotNull { folder ->
                val path = File(folder, "meta.json")
                if (!path.isFile) {
                    return@mapNotNull null
                }
                try {
                    fromJson(JSONObject(path.readText(Charsets.UTF_8)))
                } catch (_: Exception) {
                    null
                }
            }
            ?.sortedByDescending { it.createdAt }
            ?: emptyList()
    }

    fun deleteClip(clipId: String) {
        clipDir(clipId).deleteRecursively()
    }

    fun prepareReanalyze(clipId: String): ClipMeta {
        val meta = loadMeta(clipId).copy(
            status = ClipStatus.PENDING,
            error = null,
            seedBox = null,
            seeds = emptyList(),
            playStartMs = 0,
            playEndMs = null,
        )
        saveMeta(meta)
        analysisFile(clipId).delete()
        stageReportFile(clipId).delete()
        frameFeedbackFile(clipId).delete()
        return meta
    }

    fun newClipId(): String = UUID.randomUUID().toString()

    companion object {
        const val MAX_MS = 120_000L

        fun displayNameNow(now: Date = Date()): String {
            val fmt = SimpleDateFormat("MM/dd/yyyy-HH:mm", Locale.US)
            fmt.timeZone = TimeZone.getDefault()
            return fmt.format(now)
        }

        fun createdAtIso(now: Date = Date()): String {
            val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.US)
            fmt.timeZone = TimeZone.getDefault()
            return fmt.format(now)
        }

        fun formatDurationMs(ms: Int, image: Boolean = false): String {
            if (image) {
                return "--"
            }
            val total = (ms / 1000).coerceAtLeast(0)
            val m = total / 60
            val s = total % 60
            return "$m:${s.toString().padStart(2, '0')}"
        }

        fun playWindowMs(meta: ClipMeta): Pair<Long, Long> {
            val start = meta.playStartMs.toLong().coerceAtLeast(0L)
            val end = meta.playEndMs?.toLong()
                ?: if (meta.durationMs > 0) meta.durationMs.toLong() else Long.MAX_VALUE / 4
            return start to if (end <= start) start + 1 else end
        }

        fun statusKey(status: ClipStatus): String = when (status) {
            ClipStatus.PENDING -> "Pending"
            ClipStatus.PROCESSING -> "Processing"
            ClipStatus.DONE -> "Done"
        }

        fun toJson(meta: ClipMeta): JSONObject {
            val obj = JSONObject()
            obj.put("clip_id", meta.clipId)
            obj.put("created_at", meta.createdAt)
            obj.put("display_name", meta.displayName)
            obj.put("duration_ms", meta.durationMs)
            obj.put("width", meta.width)
            obj.put("height", meta.height)
            obj.put("fps", meta.fps)
            obj.put("kind", meta.kind.name.lowercase(Locale.US))
            obj.put("status", meta.status.name.lowercase(Locale.US))
            if (meta.error == null) {
                obj.put("error", JSONObject.NULL)
            } else {
                obj.put("error", meta.error)
            }
            if (meta.seedBox == null) {
                obj.put("seed_box", JSONObject.NULL)
            } else {
                obj.put("seed_box", JSONArray(meta.seedBox.toList()))
            }
            val seeds = JSONArray()
            for (seed in meta.seeds) {
                seeds.put(
                    JSONObject()
                        .put("t_ms", seed.tMs)
                        .put("box", JSONArray(seed.box.toList())),
                )
            }
            obj.put("seeds", seeds)
            obj.put("play_start_ms", meta.playStartMs)
            if (meta.playEndMs == null) {
                obj.put("play_end_ms", JSONObject.NULL)
            } else {
                obj.put("play_end_ms", meta.playEndMs)
            }
            if (meta.athleteKey == null) {
                obj.put("athlete_key", JSONObject.NULL)
            } else {
                obj.put("athlete_key", meta.athleteKey)
            }
            if (meta.athlete == null) {
                obj.put("athlete", JSONObject.NULL)
            } else {
                obj.put("athlete", meta.athlete.toJson())
            }
            if (meta.scene == null) {
                obj.put("scene", JSONObject.NULL)
            } else {
                obj.put("scene", meta.scene.toJson())
            }
            return obj
        }

        fun fromJson(obj: JSONObject): ClipMeta {
            val seedArr = obj.optJSONArray("seed_box")
            val seedsArr = obj.optJSONArray("seeds")
            val seeds = ArrayList<SeedMark>()
            if (seedsArr != null) {
                for (i in 0 until seedsArr.length()) {
                    val item = seedsArr.optJSONObject(i) ?: continue
                    val box = item.optJSONArray("box")?.let { NormBox.fromArray(it) } ?: continue
                    seeds.add(SeedMark(item.optDouble("t_ms"), box))
                }
            }
            return ClipMeta(
                clipId = obj.getString("clip_id"),
                createdAt = obj.optString("created_at"),
                displayName = obj.optString("display_name"),
                durationMs = obj.optInt("duration_ms"),
                width = obj.optInt("width", 1).coerceAtLeast(1),
                height = obj.optInt("height", 1).coerceAtLeast(1),
                fps = obj.optDouble("fps", 30.0),
                kind = if (obj.optString("kind") == "image") ClipKind.IMAGE else ClipKind.VIDEO,
                status = when (obj.optString("status")) {
                    "processing" -> ClipStatus.PROCESSING
                    "done" -> ClipStatus.DONE
                    else -> ClipStatus.PENDING
                },
                error = if (obj.isNull("error")) null else obj.optString("error").ifBlank { null },
                seedBox = seedArr?.let { NormBox.fromArray(it) },
                seeds = seeds,
                playStartMs = obj.optDouble("play_start_ms", 0.0).toInt().coerceAtLeast(0),
                playEndMs = if (obj.isNull("play_end_ms")) {
                    null
                } else {
                    obj.optDouble("play_end_ms").toInt().coerceAtLeast(0)
                },
                athleteKey = if (obj.has("athlete_key") && !obj.isNull("athlete_key")) {
                    obj.optString("athlete_key").ifBlank { null }
                } else {
                    null
                },
                athlete = obj.optJSONObject("athlete")?.let { AthleteProfile.fromJson(it) },
                scene = SceneContext.fromJson(obj.optJSONObject("scene")),
            )
        }
    }
}
